crates/core/src/node/network_bridge/p2p_protoc.rs

```diff
@@ -1499,60 +1885,41 @@ impl P2pConnManager {
 trait ConnectResultSender {
     fn send_result(
         &mut self,
-        result: Result<(PeerId, Option<usize>), HandshakeError>,
-    ) -> Pin<Box<dyn Future<Output = Result<(), HandshakeError>> + Send + '_>>;
-}
-
-impl ConnectResultSender for Option<oneshot::Sender<Result<PeerId, HandshakeError>>> {
-    fn send_result(
-        &mut self,
-        result: Result<(PeerId, Option<usize>), HandshakeError>,
-    ) -> Pin<Box<dyn Future<Output = Result<(), HandshakeError>> + Send + '_>> {
-        async move {
-            self.take()
-                .expect("always set")
-                .send(result.map(|(id, _)| id))
-                .map_err(|_| HandshakeError::ChannelClosed)?;
-            Ok(())
-        }
-        .boxed()
-    }
+        result: Result<(PeerId, Option<usize>), ()>,
+    ) -> Pin<Box<dyn Future<Output = Result<(), ()>> + Send + '_>>;
 }
 
 impl ConnectResultSender for mpsc::Sender<Result<(PeerId, Option<usize>), ()>> {
     fn send_result(
         &mut self,
-        result: Result<(PeerId, Option<usize>), HandshakeError>,
-    ) -> Pin<Box<dyn Future<Output = Result<(), HandshakeError>> + Send + '_>> {
-        async move {
-            self.send(result.map_err(|_| ()))
-                .await
-                .map_err(|_| HandshakeError::ChannelClosed)
-        }
-        .boxed()
+        result: Result<(PeerId, Option<usize>), ()>,
+    ) -> Pin<Box<dyn Future<Output = Result<(), ()>> + Send + '_>> {
+        async move { self.send(result).await.map_err(|_| ()) }.boxed()
     }
 }
 
 struct EventListenerState {
+    outbound_handler: OutboundConnectionHandler,
     // Note: peer_connections has been moved out to allow separate borrowing by the stream
     pending_from_executor: HashSet<Transaction>,
     // FIXME: we are potentially leaving trash here when transacrions are completed
     tx_to_client: HashMap<Transaction, HashSet<ClientId>>,
     client_waiting_transaction: Vec<(WaitingTransaction, HashSet<ClientId>)>,
-    transient_conn: HashMap<Transaction, SocketAddr>,
-    awaiting_connection: HashMap<SocketAddr, Box<dyn ConnectResultSender>>,
+    awaiting_connection: HashMap<SocketAddr, Vec<Box<dyn ConnectResultSender>>>,
+    awaiting_connection_txs: HashMap<SocketAddr, Vec<Transaction>>,
```

Comment:
why do we await from multiple transactions from a connection which is not yet complete? shouldnt it be a single connection transaction?

Reply:
We now keep a Vec<ConnectResultSender> per socket because multiple connect transactions can target the same remote peer concurrently (for example, a retrying JoinerState plus a courtesy dial from the maintenance loop). Previously the last transaction to register would overwrite the sender in awaiting_connection, so only one caller ever learned that the transport either connected or failed. By queuing the senders we let the single transport attempt fan the result out to every transaction that was waiting on that socket; once the connection completes we drain that vector so each ConnectOp sees a completion and they can advance (or retry) independently.